src/Uart/Uart.c

```diff
@@ -0,0 +1,427 @@
+
+static uint8_t rxFifoMemoryBlock[UART_FIFO_MEMORY_SIZE] = { 0 };
+
+static uint8_t txFifoMemoryBlock[UART_FIFO_MEMORY_SIZE] = { 0 };
+
+static Uart_ErrorFlags errorFlags[UART_INTERFACES_MAX];
+
+static uint32_t sentBytesCounter[UART_INTERFACES_MAX];
+
+static rtems_interrupt_entry uartIrqEntry[UART_INTERFACES_MAX];
+
+static inline UartRegisters_t
+addressBase(Uart_Id id)
+{
+    switch(id) {
+        case Uart_Id_0:
+            return (UartRegisters_t)Uart0_address;
+        case Uart_Id_1:
+            return (UartRegisters_t)Uart1_address;
+#ifndef RTEMS_SIS
+        case Uart_Id_2:
+            return (UartRegisters_t)Uart2_address;
+        case Uart_Id_3:
+            return (UartRegisters_t)Uart3_address;
+        case Uart_Id_4:
+            return (UartRegisters_t)Uart4_address;
+        case Uart_Id_5:
+            return (UartRegisters_t)Uart5_address;
+#endif
+        default:
+            return (UartRegisters_t)Error_address;
+    }
+}
+
+static inline Uart_BaudRate
+getBaudRate(const Uart* const uart)
+{
+    uint32_t freq = rtems_clock_get_ticks_per_second();
+    uint32_t baud = freq / (8 * ((uart->reg->clkscl << 8) + 1));
+    switch(baud) {
+        case 300:
+        case 600:
+        case 1200:
+        case 1800:
+        case 2400:
+        case 4800:
+        case 9600:
+        case 19200:
+        case 28800:
+        case 38400:
+        case 57600:
+        case 76800:
+        case 115200:
+            return (Uart_BaudRate)baud;
+        default:
+            return Error_BaudRate;
+    }
+}
+
+static inline Uart_irqNum
+irqNumber(Uart_Id id)
+{
+    switch(id) {
+        case Uart_Id_0:
+            return Uart0_irqNum;
+        case Uart_Id_1:
+            return Uart1_irqNum;
+#ifndef RTEMS_SIS
+        case Uart_Id_2:
+            return Uart2_irqNum;
+        case Uart_Id_3:
+            return Uart3_irqNum;
+        case Uart_Id_4:
+            return Uart4_irqNum;
+        case Uart_Id_5:
+            return Uart5_irqNum;
+#endif
+        default:
+            return Error_irqNum;
+    }
+}
+
+static inline void
+emptyCallback(void* arg)
+{
+    (void)arg;
+}
+
+static Uart_TxHandler defaultTxHandler = { .callback = emptyCallback,
+                                           .arg = 0 };
+
+static Uart_RxHandler defaultRxHandler = { .lengthCallback = emptyCallback,
+                                           .characterCallback = emptyCallback,
+                                           .lengthArg = 0,
+                                           .characterArg = 0,
+                                           .targetCharacter = '\0',
+                                           .targetLength = 0 };
+
+static Uart_ErrorHandler defaultErrorHandler = { .callback = emptyCallback,
+                                                 .arg = 0 };
+
+void
+Uart_setConfig(Uart* const uart, const Uart_Config* const config)
+{
+    uart->config.isTxEnabled = config->isTxEnabled;
```

Comment:
Are we duplicating the contents of the registers in the Uart structure?

Reply:
my intention is to use Uart_Config structure as desired configuration of Uart rather than just representation of state of internal registers, I can change this after discussion

Reply:
config in Uart struct in SAMV71 BSP seems to be due to the fact that SAMV71 registers do not expose the entire config for reading. For Leon3, this is redundancy, please remove.